src/pool/srv_pool.c

```diff
@@ -1458,6 +1458,47 @@ resume_event_handling(struct pool_svc *svc)
 	ABT_mutex_unlock(events->pse_mutex);
 }
 
+/*
+ * Restart rebuild if the rank is UPIN in pool map and is in rebuilding.
+ *
+ * This function only used when PS leader gets CRT_EVT_ALIVE event of engine \a rank,
+ * if that rank is UPIN in pool map and with unfinished rebuilding should be massive
+ * failure case -
+ * 1. some engines down and triggered rebuild.
+ * 2. the engine \a rank participated the rebuild, not finished yet, it became down again,
+ *    the #failures exceeds pool RF and will not change pool map.
+ * 3. That engine restarted by administrator.
+ *
+ * In that case should recover the rebuild task on engine \a rank, to simplify it now just
+ * abort and retry the global rebuild task.
+ */
+static void
+pool_restart_rebuild_if_rank_wip(struct ds_pool *pool, d_rank_t rank)
+{
+	struct pool_domain	*dom;
+
+	ABT_rwlock_rdlock(pool->sp_lock);
```

Comment:
The caller of `pool_restart_rebuild_if_rank_wip` already holds this lock.

Reply:
Ah, right, I remember I checked the lock usage but maybe due to some mistake.
BTW, ABT_rwlock_rdlock seems can reentry, as I tested the code path manually.

Reply:
Yeah, I assumed that must be the case, the ability to be reentered. It could be just a lucky coincidence, because Argobots's documentation does not promise `ABT_rwlock_rdlock` can be used recursively.